File: backend/src/tipos/clases/tipos_de_metodos/methods.js
"use strict"

class Base {}  // Clase abstracta
    constructor() {
        if (this.constructor === Base) {
            var msg = "La clase asbtracta Base " +
                "no puede ser instanciada directamente."
            throw new TypeError(msg);
        }
              // Método abstracto
        if (this.metodo_abstracto === undefined) {
            var msg = "Debes declarar el método" +
                + "methodo_abstracto() en" + this.constructor.name;
            throw new TypeError(msg);
        }
    }
;

class Ejemplo extends Base {

	// Constructor
	constructor(atributo) {
		this.atributo = atributo;
	}

	normal() {
		console.log("\nSoy un método normal");
		console.log(this);
	}

	// Getter
	get atributo() {
		return this._atributo;
	}

	// Setter
	set atributo(value){
		if (value.length < 4) {
			console.log("Atributo muy corto");
			return
		}
		this._atributo = value;
	}

	// Estático
	static estatico() {
		console.log("\nSoy un método estático");
		console.log("Debes llamarme desde la clase");
		console.log("En Python me llamarías método de clase");
	}

	metodo_abstracto(){
		console.log("Mira en la clase padre.")
		console.log("Cámbiame el nombre en Ejemplo y verás lo que pasa.")
	}

}

// Instanciar
var ej = new Ejemplo("a");   // Atributo muy corto

var ej = new Ejemplo("Soy un atributo");

// Inspeccionar clase   -->  [ 'constructor', 'normal', ... ]
console.log(Object.getOwnPropertyNames(Ejemplo.prototype));

// Acceso a atributos
console.log(ej.atributo);  // Soy un atributo

// Método normal
ej.normal();  // Soy un método normal

// Método estático
Ejemplo.estatico();  // Soy un método estático
                     // Debes llamarme desde la clase
                     // En Python me llamarías método de clase

// Ḿétodo abstracto
ej.metodo_abstracto() // Mira en la clase padre.
                      // Cámbiame el nombre en Ejemplo y verás lo que pasa.

File: backend/src/tipos/clases/tipos_de_metodos/methods.rb
#!/usr/bin/ruby

class Base;
    def initialize()
        if !defined? self.abstracto()
            raise TypeError, "Debes declarar el método " +
                "abstracto() en #{self.class.name}"
        end
    end
end;

class Ejemplo < Base  # Herencia
    @@propiedad = "Soy una propiedad"

    def self.propiedad
    	@@propiedad
    end

    # Constructor
    def initialize
    	@_atributo = "Soy un atributo"
    	@_atributo_leible = "Soy un atributo leible"
    	@_atributo_sobrescribible = "Soy un atributo sobreescribible"
    end

	# Atributo leible y editable
	attr_accessor :_atributo

	attr_reader :_atributo_leible # Sólo leíble

	attr_writer :_atributo_sobrescribible # Sólo editable

    # Método público por defecto
	def normal; puts "Soy un método normal"; end;

	# Método de clase
	def self.de_clase
		return Ejemplo.new
	end

    def abstracto
        puts "Mira en la clase padre."
        puts "Cambiame el nombre en #{self.class.name} y verás lo que pasa."
    end

    # Método para obtener las clases hijas
    def self.descendants
        ObjectSpace.each_object(Class).select { |klass| klass < self }
    end


	protected
	    def protegido
	    	"Método protegido"
	    end

    private
        def privado;
        	"Método privado"
        end
end

# Constructor
ej = Ejemplo.new

# Acceso a propiedades
puts Ejemplo.propiedad  # Soy una propiedad

puts ej._atributo # Soy un atributo

#puts ej._atributo_sobrescribible  # ERROR --> undefined method `_atributo_sobrescribible

# Método normal
ej.normal()  # Soy un método normal

# Método de clase
puts Ejemplo.de_clase  # #<Ejemplo:<dirección_en_memoria>>

# Método abstracto
puts ej.abstracto  # # Mira en la clase padre.
                   # Cambiame el nombre en Ejemplo y verás lo que pasa.

# Método protegido
#puts ej.protegido  # ERROR --> protected method `protegido' called

# Método privado
#puts ej.privado  # ERROR --> private method `privado' called

# ====================================

# Obtener las subclases
ej.descendants()


=begin

#####     RANGOS DE ACCESO     #####

- Los métodos públicos pueden ser invocados
por cualquiera, no tienen control de acceso.
Los métodos son públicos por defecto, excepto
initialize(), que siempre es privado.
- Los métodos protegidos sólo pueden ser
invocados por objetos de la clase que los define
o sus subclases. Su acceso se mantiene dentro
de la familia.
- Los métodos privados sólo pueden ser llamados
en la clase que los definen y dentro de los
descendientes directos.

=end
